contributors/design-proposals/architecture/scope.md

```diff
@@ -0,0 +1,342 @@
+  there many reasonable alternative approaches and solutions to the
+  problem?
+
+* Is it an area where we want to foster exploration and innovation in
+  the ecosystem?
+
+* Has the ecosystem produced adequate solutions on its own? For
+  instance, have ecosystem projects taken on requirements of the
+  Kubernetes project, if needed? Example: etcd3 added a number of features
+  and other improvements to benefit Kubernetes, so the project didn't
+  need to launch a separate storage effort.
+
+* Is there an acceptable home for the recommended ecosystem solution(s)?
+  Example: the [CNCF Sandbox](https://github.com/cncf/toc/blob/master/process/sandbox.md) is one possible home
+
+* Has the functionality been provided by the project/release/component
+  historically?
+
+## Technical scope details and rationale
+
+### Containerized workload execution and management
+
+Including:
+* common general categories of workloads, such as stateless, stateful, batch, and cluster services
+* provisioning, allocation, accessing, and managing compute, storage, and network resources on behalf of the workloads, and enforcement of security policies on those resources
+* workload prioritization, capacity assessment, placement, and relocation (aka scheduling)
+* graceful workload eviction
+* local container image caching
+* configuration and secret distribution
+* manual and automatic horizontal and vertical scaling
+* deployment, progressive (aka rolling) upgrades, and downgrades
+* self-healing
+* exposing container logs, status, health, and resource usage metrics for collection
+
+### Service discovery, load balancing, and routing
+
+Including:
+* endpoint tracking and discovery, including pod and non-pod endpoints
+* the most common L4 and L7 Internet protocols (TCP, UDP, SCTP, HTTP, HTTPS)
+* intra-cluster DNS configuration and serving
+* external DNS configuration
+* accessing external services (e.g., imported services, Open Service Broker)
+* exposing traffic latency, throughput, and status metrics for collection
+* access authorization
+
+### Workload identity propagation and authentication
+
+Including:
+* internal identity (e.g., SPIFFE support)
+* external identity (e.g., TLS certificate management)
+
+### Declarative resource management platform
+
+Including:
+* CRUD API operations and behaviors, diff, patch, dry run, watch
+* declarative updates (apply)
+* resource type definition, registration, discovery, documentation, and validation mechanisms
+* pluggable authentication, authorization, admission (API-level policy enforcement), and audit-logging mechanisms
+* Namespace (resource scoping primitive) lifecycle
+* resource instance persistence and garbage collection
+* asynchronous event reporting
+* API producer SDK
+* API client SDK / libraries in widely used languages
+* dynamic, resource-oriented CLI, as a reference implementation for interacting with the API and basic tool for declarative and imperative management
```

Comment:
> can be used as a final step by higher level tooling to avoid recreating common code like apply or create

I would prefer that the final step for higher level tooling be baked into client libraries or the API. This would allow many tools to use those features without having shell out to kubectl or deal with making sure it's installed at a right dependency version.

Why can't these things be baked into the API and client libs?

Reply:
> higher level CLI tooling that allows common day to day workflows of applying core API types like config and secrets to the cluster repeatedly

Whose higher level workflows and how to we figure out what those are? Since most of the Kubernetes users aren't in meetings or involved in the community how do we determine the workflows to chase after?

If the base level features were built into client libraries and APIs people could more easily create their own workflows as needed. The k8s client would not need to try and figure out the workflows to chase.

Reply:
These tools are, and we do, and we have been for most of the last two years.

But there's a big difference between "some go code that lives in a library" and "go code that is used to solve a problem for a user right now".  I'm not interested in the generic "we could build this in so many ways", I'm much more focused on reducing day to day user stress within Kubernetes.  All libraries need one major consumer giving them purpose, and our declarative resource management tool kubectl is the major consumer of those libraries.

Reply:
kubectl already implements validation using the OpenAPI spec, which https://github.com/garethr/kubeval also uses. 

Server-side validation is still something I'd like done (https://github.com/kubernetes/kubernetes/issues/5889), but has additional challenges (backward compatibility, version skew, how to return warnings, how to make error messages understandable and tool-friendly, ...). And we decided diff and server-side dry run were more urgent.